src/components/StateSelector.tsx

```diff
@@ -0,0 +1,100 @@
+import type {MaybePhraseKey} from '@libs/Localize';
+import Navigation from '@libs/Navigation/Navigation';
+import ROUTES from '@src/ROUTES';
+import FormHelpMessage from './FormHelpMessage';
+import type {MenuItemProps} from './MenuItem';
+import MenuItemWithTopDescription from './MenuItemWithTopDescription';
+
+type State = keyof typeof COMMON_CONST.STATES;
+
+type StateSelectorProps = {
+    /** Form error text. e.g when no country is selected */
+    errorText?: MaybePhraseKey;
+
+    /** Current selected state  */
+    value?: State;
+
+    /** Callback to call when the input changes */
+    onInputChange?: (value: string) => void;
+
+    /** Label to display on field */
+    label?: string;
+
+    /** Any additional styles to apply */
+    wrapperStyle?: MenuItemProps['wrapperStyle'];
+
+    /** whether to use state from url, for cases when url value is passed from parent */
+    shouldUseStateFromUrl?: boolean;
+};
+
+function StateSelector({errorText, shouldUseStateFromUrl = true, value: stateCode, label, onInputChange, wrapperStyle}: StateSelectorProps, ref: ForwardedRef<View>) {
+    const styles = useThemeStyles();
+    const {translate} = useLocalize();
+    const stateFromUrl = useGeographicalStateFromRoute();
+    const [stateToDisplay, setStateToDisplay] = useState<State | ''>('');
+
+    useEffect(() => {
+        if (!shouldUseStateFromUrl || !stateFromUrl) {
+            return;
+        }
+
+        // This will cause the form to revalidate and remove any error related to country name
+        if (onInputChange) {
+            onInputChange(stateFromUrl);
+        }
+        setStateToDisplay(stateFromUrl);
+
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [stateFromUrl, shouldUseStateFromUrl]);
+
+    useEffect(() => {
+        if (!stateCode) {
+            return;
+        }
+
+        if (onInputChange) {
+            // This will cause the form to revalidate and remove any error related to state name
+            onInputChange(stateCode);
+        }
+        setStateToDisplay(stateCode);
+
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [stateCode]);
+
+    const title = stateToDisplay && Object.keys(COMMON_CONST.STATES).includes(stateToDisplay) ? translate(`allStates.${stateToDisplay}.stateName`) : '';
+    const descStyle = title.length === 0 ? styles.textNormal : null;
+
+    return (
+        <View>
+            <MenuItemWithTopDescription
+                descriptionTextStyle={descStyle}
+                ref={ref}
+                shouldShowRightIcon
+                title={title}
+                // Label can be an empty string
+                // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+                description={label || translate('common.state')}
+                onPress={() => {
+                    const activeRoute = Navigation.getActiveRoute();
+                    // @ts-expect-error Navigation.navigate does take a param
```

Comment:
I had to add `// @ts-expect-error Navigation.navigate does take a param` at places because TS was giving me error. But i dont think this problem is specific to my code. What do you say?

Reply:
cc @blazejkustra @fabioh8010 for visibility

Reply:
@ygshbht I've checked here and it's specific to your code. It's happening for two reasons:

1. You are making `SETTINGS_ADDRESS_STATE`'s `getRoute` function return a `string` [here](https://github.com/Expensify/App/blob/77f236c7822570bf9f31db1d2563c852033291fa/src/ROUTES.ts#L136), but [we have a condition](https://github.com/Expensify/App/blob/77f236c7822570bf9f31db1d2563c852033291fa/src/ROUTES.ts#L554) to make the `Route` type be `never` if we have a route that returns `string`. This is to ensure all routes return only constants, I think you can solve it with this code:
    ```ts
        SETTINGS_ADDRESS_STATE: {
            route: 'settings/profile/address/state',
            getRoute: (state: string, backTo?: string, label?: string) =>
                `${getUrlWithBackToParam(`settings/profile/address/state?state=${state}`, backTo)}${label ? (`&label=${encodeURIComponent(label)}` as const) : ''}` as const,
        },
    ```
    The adjusted code will make `getRoute` return all the possible route strings as constant values instead of just `string`, thus eliminating the problem.
2. After doing the first fix, if you remove the `@ts-expect-error` comment you will have another error:
    <img width="1010" alt="Screenshot 2024-02-20 at 09 42 28" src="https://github.com/Expensify/App/assets/20051562/ced15857-a096-456a-8d9d-e5735abf53a9">
    Since `stateCode` can be `undefined` and your route function doesn't accept that, to fix you can do `stateCode ?? ''` if it's safe enough, or change `getRoute` signature to accept `undefined` values there and handle it, or put a condition to only call the navigation if `stateCode` is defined.
    ```ts
                    onPress={() => {
                        if (!stateCode) {
                            return;
                        }
    
                        const activeRoute = Navigation.getActiveRoute();
                        Navigation.navigate(ROUTES.SETTINGS_ADDRESS_STATE.getRoute(stateCode, activeRoute, label));
                    }}
    ```
    The best solution will depends on how this fix/feature works.

Reply:
Thanks, I've made the change